File: 8/8.13.swift
protocol DiscountBase {
    var price: Int { get }
    // 準拠先で違いのある割引金額をパラメータとして用意した
    var discountCharge: Int { get }
    func getDiscountedPrice() -> Int
}

extension DiscountBase {
    // 割引き計算を共通化した
    func getDiscountedPrice() -> Int {
        max(0, price - discountCharge)
    }
}

// 通常の割引き
class RegularDiscount: DiscountBase {
    private(set) var price: Int
    // [BAD]
    // このクラスだけ見てもdiscountChargeが何のためにあるのか分からない
    private(set) var discountCharge: Int = 300
    init(price: Int) {
        self.price = price
    }
}

// 夏の割引き
class SummerDiscount: DiscountBase {
    private(set) var price: Int
    // [BAD]
    // このクラスだけ見てもdiscountChargeが何のためにあるのか分からない
    private(set) var discountCharge: Int = 400
    init(price: Int) {
        self.price = price
    }
}

File: 8/8.13.dart
import 'dart:math';

// Dartでのabstract classを使った宣言
abstract class DiscountBase {
  int get price;
  // 準拠先で違いのある割引金額をパラメータとして用意した
  int get discountCharge;
  // 割引き計算を共通化した
  int getDiscountedPrice() => max(0, price - discountCharge);
}

// 通常の割引き
class RegularDiscount extends DiscountBase {
  @override
  final int price;
  // [BAD]
  // このクラスだけ見てもdiscountChargeが何のためにあるのか分からない
  @override
  final int discountCharge = 300;

  RegularDiscount(this.price);
}

// 夏の割引き
class SummerDiscount extends DiscountBase {
  @override
  final int price;
  // [BAD]
  // このクラスだけ見てもdiscountChargeが何のためにあるのか分からない
  @override
  final int discountCharge = 400;

  SummerDiscount(this.price);
}

void main() {
  final regular = RegularDiscount(1000);
  final summer = SummerDiscount(1000);

  print(regular.getDiscountedPrice());  // 700
  print(summer.getDiscountedPrice());   // 600
}
